File: src/main/java/no/novari/integration/IntegrationMappingService.java
package no.novari.integration;

import no.novari.integration.model.dtos.IntegrationDto;
import no.novari.integration.model.dtos.IntegrationPostDto;
import no.novari.integration.model.entities.Integration;
import org.springframework.stereotype.Service;

import java.util.Collection;

@Service
public class IntegrationMappingService {

    public Integration toIntegration(IntegrationPostDto integrationPostDto) {
        return Integration
                .builder()
                .sourceApplicationId(integrationPostDto.getSourceApplicationId())
                .sourceApplicationIntegrationId(integrationPostDto.getSourceApplicationIntegrationId())
                .destination(integrationPostDto.getDestination())
                .state(Integration.State.DEACTIVATED)
                .build();
    }

    public Collection<IntegrationDto> toDtos(Collection<Integration> integrations) {
        return integrations.stream().map(this::toDto).toList();
    }

    public IntegrationDto toDto(Integration integration) {
        return IntegrationDto
                .builder()
                .id(integration.getId())
                .sourceApplicationId(integration.getSourceApplicationId())
                .sourceApplicationIntegrationId(integration.getSourceApplicationIntegrationId())
                .destination(integration.getDestination())
                .state(integration.getState())
                .activeConfigurationId(integration.getActiveConfigurationId())
                .build();
    }

}

File: src/main/kotlin/no/novari/flyt/integration/application/IntegrationMappingService.kt
package no.novari.flyt.integration.application

import no.novari.flyt.audit.actor.ActorDisplayResolver
import no.novari.flyt.integration.api.dto.IntegrationDto
import no.novari.flyt.integration.api.dto.IntegrationPostDto
import no.novari.flyt.integration.persistence.entity.Integration
import org.springframework.stereotype.Service

@Service
class IntegrationMappingService(
    private val actorDisplayResolver: ActorDisplayResolver,
) {
    fun toIntegration(integrationPostDto: IntegrationPostDto): Integration {
        return Integration(
            sourceApplicationId = requireNotNull(integrationPostDto.sourceApplicationId),
            sourceApplicationIntegrationId = requireNotNull(integrationPostDto.sourceApplicationIntegrationId),
            destination = requireNotNull(integrationPostDto.destination),
            state = Integration.State.DEACTIVATED,
        )
    }

    fun toDtos(integrations: Collection<Integration>): List<IntegrationDto> {
        val createdByDisplays = actorDisplayResolver.resolveAll(integrations.map { it.createdBy })
        val lastModifiedByDisplays = actorDisplayResolver.resolveAll(integrations.map { it.lastModifiedBy })
        return integrations.map {
            toDto(it, createdByDisplays[it.createdBy], lastModifiedByDisplays[it.lastModifiedBy])
        }
    }

    fun toDto(integration: Integration): IntegrationDto {
        return toDto(
            integration,
            actorDisplayResolver.resolve(integration.createdBy),
            actorDisplayResolver.resolve(integration.lastModifiedBy),
        )
    }

    private fun toDto(
        integration: Integration,
        createdByDisplay: String?,
        lastModifiedByDisplay: String?,
    ): IntegrationDto {
        return IntegrationDto(
            id = requireNotNull(integration.id),
            sourceApplicationId = requireNotNull(integration.sourceApplicationId),
            sourceApplicationIntegrationId = requireNotNull(integration.sourceApplicationIntegrationId),
            destination = requireNotNull(integration.destination),
            state = requireNotNull(integration.state),
            activeConfigurationId = integration.activeConfigurationId,
            createdAt = integration.createdAt,
            createdBy = createdByDisplay,
            createdByActor = integration.createdBy,
            lastModifiedAt = integration.lastModifiedAt,
            lastModifiedBy = lastModifiedByDisplay,
            lastModifiedByActor = integration.lastModifiedBy,
        )
    }
}
